src/neo/SmartContract/JsonSerializer.cs

```diff
@@ -16,49 +17,89 @@ public static class JsonSerializer
+        {
+            JObject returnValue;
             switch (item)
             {
                 case Array array:
                     {
-                        return array.Select(p => Serialize(p)).ToArray();
+                        var list = new List<JObject>();
+                        foreach (var arrayItem in array)
+                        {
+                            int usedBytes = 0;
+                            list.Add(Serialize(arrayItem, out usedBytes, preExistingSize, maxSize));
+                            preExistingSize += usedBytes;
+                        }
+                        returnValue = list.ToArray();
+                        break;
                     }
                 case ByteArray buffer:
                     {
-                        return Convert.ToBase64String(buffer.GetSpan());
+                        returnValue = Convert.ToBase64String(buffer.GetSpan());
+                        break;
                     }
                 case Integer num:
                     {
                         var integer = num.GetBigInteger();
                         if (integer > JNumber.MAX_SAFE_INTEGER || integer < JNumber.MIN_SAFE_INTEGER)
-                            return integer.ToString();
-                        return (double)num.GetBigInteger();
+                        {
+                            returnValue = integer.ToString();
+                        }
+                        else
+                        {
+                            returnValue = (double)num.GetBigInteger();
+                        }
+                        break;
                     }
                 case Boolean boolean:
                     {
-                        return boolean.ToBoolean();
+                        returnValue = boolean.ToBoolean();
+                        break;
                     }
                 case Map map:
                     {
                         var ret = new JObject();
-
                         foreach (var entry in map)
                         {
+                            int usedBytes = 0;
                             var key = entry.Key.GetString();
-                            var value = Serialize(entry.Value);
-
+                            var value = Serialize(entry.Value, out usedBytes, preExistingSize, maxSize);
                             ret[key] = value;
+                            preExistingSize += usedBytes;
                         }
 
-                        return ret;
+                        returnValue = ret;
+                        break;
                     }
                 case Null _:
                     {
-                        return JObject.Null;
+                        returnValue = JObject.Null;
+                        break;
                     }
                 default: throw new FormatException();
             }
+
+            consumedBytes = returnValue.ToByteArray(false).Length;
```

Comment:
Do you guys think that there is a better way to do this check? I don't feel comfortable with having to call this several times.
Do you think we could cache the serialization results? Can that save us time and increase performance?

Reply:
This is realy not good.

Reply:
@erikzhang  maybe we need to store each object size. My only concern is serializing a big object before checking the size. If there are no security issues, I think we can forget this subject.

Reply:
I improved it in the original PR.